### ISSUE TITLE
Support for Hyperboost

### ISSUE THREAD

**Issue by Yatoom** [automl/SMAC3#915]
Hi, I have been working on a surrogate model (Hyperboost) based on gradient boosting. This seems to outperform SMAC's random forest in most cases, while the training and querying of the surrogate model takes less time.

I think it could be intersting to built this into SMAC.

# Concept
The basic idea is based on using [virtual ensembles](https://towardsdatascience.com/tutorial-uncertainty-estimation-with-catboost-255805ff217e) to retrieve uncertainty estimation. This functionality is built into CatBoost, which comes with fast training.  You can create an EPM out of this that SMAC can use. A simplified version of the code would look like this:

```python
def _train(X, Y):
  # This configuration was found after trying many different settings and running the optimizer on 
  # 31 datasets using a small random forest target algorithm.
  self.model = CatBoostRegressor(iterations=100, loss_function="RMSEWithUncertainty", posterior_sampling=False,
                                 verbose=False, random_seed=0, learning_rate=1.0, random_strength=0, 
                                 l2_leaf_reg=1)
  self.model.fit(X, Y)

def _predict(self, X):
  pred = self.catboost.predict(X)
  ensemble_preds = self.catboost.virtual_ensembles_predict(X, prediction_type="TotalUncertainty",
                                                           virtual_ensembles_count=20)
  knowledge = ensemble_preds[:, 1]

  # The knowledge uncertainty returned by virtual ensembles is much lower than using actual ensembles.
  # Scaling them up will approximately give numbers in the same size as the real ensembles.
  return pred[:, 0], knowledge ** 0.3 
```

# Benchmarking
I have created a benchmarking tool ([Hyperbench](https://github.com/Yatoom/hyperbench)) to test the final result on SMAC 1.4.

- The configuration for this benchmark can be found [here](https://github.com/Yatoom/hyperbench/blob/main/run_benchmark.py)
- The SMAC configuration can be found [here](https://github.com/Yatoom/hyperbench/blob/main/hyperbench/optimizers/smac_based.py)
- The target algorithms including their config spaces can be found [here](https://github.com/Yatoom/hyperbench/tree/main/hyperbench/target_algorithms)
- The Hyperboost EPM can be found [here](https://github.com/Yatoom/hyperbench/blob/main/hyperbench/hyperboost/epm.py)

The results of the benchmark are (currently) included in the Hyperbench repository. You could clone it, install the dependencies and run `streamlit run dashboard.py` to view the results interactively. Alternatively, you could have a look at the graphs I have included below.

- The SMAC output is also stored inside the repository [here](https://github.com/Yatoom/hyperbench/tree/main/smac_output)

# Results
## Time required for EPM
![image](https://user-images.githubusercontent.com/4205641/216735082-ce65a5ed-cd74-4c57-badb-e82b37110c0f.png)

### XGBoost
![image](https://user-images.githubusercontent.com/4205641/216731708-ed0909f5-27d3-4aa4-8419-8bf9b0497d4d.png)

### Random Forest
![image](https://user-images.githubusercontent.com/4205641/216731816-3fdf0600-f7e0-430c-8fd2-d64ae8555f8c.png)

### SVM
![image](https://user-images.githubusercontent.com/4205641/216731878-9bd9770a-6081-4d43-b47e-344fc111b5a2.png)

### Stochastic gradient descent
![image](https://user-images.githubusercontent.com/4205641/216731961-e77715c5-b8ee-44e4-8c9a-77e1c667b096.png)
